File: test/nodejs/consumer-test.mjs
#!/usr/bin/env node

// Consumer verification test for @pulsyflux/broker
//
// This script simulates an external consumer installing and using the
// published npm package. It creates a temporary directory, installs
// @pulsyflux/broker via pnpm, and verifies the postinstall pipeline
// produced the expected native artifacts and that the public API is usable.
//
// NOTE: This test requires @pulsyflux/broker to be published on npm.
// It will fail until the package is available on the registry.

import { mkdtemp, writeFile, rm, access } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { execSync } from "node:child_process";

const PASS = "\x1b[32mPASS\x1b[0m";
const FAIL = "\x1b[31mFAIL\x1b[0m";
const SKIP = "\x1b[33mSKIP\x1b[0m";

let tmpDir;
let exitCode = 0;

function log(status, msg) {
  console.log(`  [${status}] ${msg}`);
}

try {
  tmpDir = await mkdtemp(join(tmpdir(), "broker-node-consumer-"));

  // 1. Write a minimal package.json that depends on @pulsyflux/broker
  const pkg = {
    name: "test-consumer",
    version: "1.0.0",
    private: true,
    type: "module",
    dependencies: {
      "@pulsyflux/broker": "latest",
    },
  };
  await writeFile(join(tmpDir, "package.json"), JSON.stringify(pkg, null, 2));

  // 2. Run pnpm install
  try {
    execSync("pnpm install", {
      cwd: tmpDir,
      stdio: "pipe",
      timeout: 120_000,
    });
    log(PASS, "pnpm install succeeded");
  } catch (err) {
    log(FAIL, `pnpm install failed: ${err.stderr?.toString().trim() || err.message}`);
    log(SKIP, "Skipping remaining checks (package may not be published yet)");
    process.exit(0);
  }

  // 3. Verify native artifacts exist
  const pkgDir = join(tmpDir, "node_modules", "@pulsyflux", "broker");
  const dllPath = join(pkgDir, ".bin", "release", "broker_lib.dll");
  const addonPath = join(pkgDir, ".bin", "release", "broker_addon.node");

  try {
    await access(dllPath);
    log(PASS, "broker_lib.dll exists");
  } catch {
    log(FAIL, `broker_lib.dll not found at ${dllPath}`);
    exitCode = 1;
  }

  try {
    await access(addonPath);
    log(PASS, "broker_addon.node exists");
  } catch {
    log(FAIL, `broker_addon.node not found at ${addonPath}`);
    exitCode = 1;
  }

  // 4. Import the package and verify Server and Client are available
  try {
    const broker = await import(join(pkgDir, ".bin", "release", "registry.mjs"));

    if (typeof broker.Server !== "function") {
      log(FAIL, "Server is not a constructor");
      exitCode = 1;
    } else {
      log(PASS, "Server class is available");
    }

    if (typeof broker.Client !== "function") {
      log(FAIL, "Client is not a constructor");
      exitCode = 1;
    } else {
      log(PASS, "Client class is available");
    }
  } catch (err) {
    log(FAIL, `Import failed: ${err.message}`);
    exitCode = 1;
  }
} finally {
  // Cleanup
  if (tmpDir) {
    await rm(tmpDir, { recursive: true, force: true }).catch(() => {});
  }
}

if (exitCode === 0) {
  console.log("\nNode.js consumer test: ALL PASSED");
} else {
  console.log("\nNode.js consumer test: SOME FAILED");
}

process.exit(exitCode);

File: test/go/consumer_test.go
package consumer_test

import (
	"os"
	"os/exec"
	"path/filepath"
	"testing"
)

// TestGoConsumerCanBuild verifies that an external Go module can import
// github.com/pulsyflux/broker and compile successfully. This test creates
// a temporary Go module that depends on the broker package, writes a
// main.go that exercises the public API (Server, Client, Publish), and
// runs `go build` to confirm it compiles.
//
// NOTE: This test requires the Go module to be published at
// github.com/pulsyflux/broker. It will fail until the module is available
// on the Go module proxy.
func TestGoConsumerCanBuild(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "broker-go-consumer-*")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	// Write go.mod for the temporary consumer module
	goMod := `module testconsumer

go 1.25

require github.com/pulsyflux/broker v1.0.0
`
	if err := os.WriteFile(filepath.Join(tmpDir, "go.mod"), []byte(goMod), 0644); err != nil {
		t.Fatalf("failed to write go.mod: %v", err)
	}

	// Write main.go that imports and uses the broker package
	mainGo := `package main

import (
	"fmt"
	"github.com/google/uuid"
	"github.com/pulsyflux/broker/broker"
)

func main() {
	// Create and start a server
	srv := broker.NewServer("127.0.0.1:0")
	if err := srv.Start(); err != nil {
		fmt.Printf("server start error: %v\n", err)
		return
	}
	defer srv.Stop()

	addr := srv.Addr()
	channelID := uuid.New()

	// Create a client and publish a message
	client, err := broker.NewClient(addr, channelID)
	if err != nil {
		fmt.Printf("client create error: %v\n", err)
		return
	}

	if err := client.Publish([]byte("hello from consumer test")); err != nil {
		fmt.Printf("publish error: %v\n", err)
		return
	}

	fmt.Println("broker consumer test: OK")
}
`
	if err := os.WriteFile(filepath.Join(tmpDir, "main.go"), []byte(mainGo), 0644); err != nil {
		t.Fatalf("failed to write main.go: %v", err)
	}

	// Run go mod tidy to resolve dependencies
	tidy := exec.Command("go", "mod", "tidy")
	tidy.Dir = tmpDir
	tidy.Stdout = os.Stdout
	tidy.Stderr = os.Stderr
	if err := tidy.Run(); err != nil {
		t.Skipf("go mod tidy failed (module may not be published yet): %v", err)
	}

	// Run go build to verify compilation
	build := exec.Command("go", "build", "-o", filepath.Join(tmpDir, "testconsumer"), ".")
	build.Dir = tmpDir
	build.Stdout = os.Stdout
	build.Stderr = os.Stderr
	if err := build.Run(); err != nil {
		t.Fatalf("go build failed: %v", err)
	}

	t.Log("Go consumer build succeeded")
}
